### PR TITLE
Collection strategies

### LINKED ISSUE #4: feature: api.Collector.register() name collision policy
Would like to be able to specify name collision policy.

"accumulate": collect() returns a dict-of-lists

"raise": raise an error on collision

"first" / "last": only accept if nothing present / overwrite always

(for first and last to make sense, a deterministic order of collect walking may be required)

### PR DESCRIPTION
**Pull request by moshez**
Fixes #4

### CODE REVIEW

File: src/gather/api.py [hunk 1]
```diff
@@ -49,6 +50,23 @@ class Collector(object):
...
+    @staticmethod
+    def all(registry, effective_name, objct):
+        myset = registry.get(effective_name, set())
+        myset.add(objct)
+        return myset
+
+    @staticmethod
+    def conflict(registry, effective_name, objct):
+        if effective_name in registry:
+            raise ValueError("Attempt to double register",
```
Comment (markrwilliams) [inline src/gather/api.py]: Should this be an exception specific to `gather` instead of `ValueError`?

File: src/gather/api.py [hunk 2]
```diff
@@ -72,13 +90,18 @@ def callback(scanner, inner_name, objct):
                 effective_name = inner_name
             else:
                 effective_name = name
-            scanner.registry[effective_name] = objct
+            try:
+                scanner.registry[effective_name] = scanner.strategy(scanner.registry,
+                                                                    effective_name,
+                                                                    objct)
+            except ValueError as exc:
```
Comment (markrwilliams) [inline src/gather/api.py]: See above.

File: src/gather/api.py [hunk 3]
```diff
@@ -87,9 +110,11 @@ def collect(self):
         def ignore_import_error(_unused):
             if not issubclass(sys.exc_info()[0], ImportError):
                 raise # pragma: no cover
-        scanner = venusian.Scanner(registry=registry, tag=self)
+        scanner = venusian.Scanner(registry=registry, tag=self, strategy=strategy)
         for module in _get_modules():
             scanner.scan(module, onerror=ignore_import_error)
+        if hasattr(scanner, 'please_raise'):
```
Comment (markrwilliams) [inline src/gather/api.py]: `hasattr` suppresses all exceptions on Python 2.7 so it's [no good](https://hynek.me/articles/hasattr/).

Why not 

```python
please_raise = getattr(scanner, 'please_raise', None)
if please_raise is not None:
    raise please_raise
```

File: src/gather/api.py [hunk 4]
```diff
@@ -49,6 +50,23 @@ class Collector(object):
     and *collect*-ing them when they need to be used.
     """
 
+    @staticmethod
+    def one_of(_registry, _effective_name, objct):
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?  Also, a `staticmethod`?  Why not a module-level function?  That would remove the need to grab `__func__` below.

File: src/gather/api.py [hunk 5]
```diff
@@ -49,6 +50,23 @@ class Collector(object):
...
+        return objct
+
+    @staticmethod
+    def all(registry, effective_name, objct):
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?

File: src/gather/api.py [hunk 6]
```diff
@@ -49,6 +50,23 @@ class Collector(object):
...
+        myset = registry.get(effective_name, set())
+        myset.add(objct)
+        return myset
+
+    @staticmethod
+    def conflict(registry, effective_name, objct):
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?  Especially what it raises!

File: src/gather/api.py [hunk 7]
```diff
@@ -72,13 +90,18 @@ def callback(scanner, inner_name, objct):
                 effective_name = inner_name
             else:
                 effective_name = name
-            scanner.registry[effective_name] = objct
+            try:
+                scanner.registry[effective_name] = scanner.strategy(scanner.registry,
+                                                                    effective_name,
+                                                                    objct)
+            except ValueError as exc:
+                scanner.please_raise = exc
```
Comment (markrwilliams) [inline src/gather/api.py]: I wonder if this and the `tag` attribute together represent enough mass to warrant a context-like object with defined attributes.  Thoughts?

File: src/gather/tests/test_commands.py [hunk 8]
```diff
@@ -31,6 +31,23 @@ def main4(args):
 def weird_function():
     pass
 
+CONFLICTING_COMMANDS = gather.Collector()
```
Comment (markrwilliams) [inline src/gather/tests/test_commands.py]: I understand `venusian` makes this relatively safe, but would it be possible to put these into the `setUp` method?  If mutation isn't a concern, sharing fixtures between tests might be.

File: src/gather/tests/test_commands.py [hunk 9]
```diff
@@ -42,14 +59,42 @@ def test_collecting(self):
...
 
     def test_cross_module_collection(self):
         collected = _helper.WEIRD_COMMANDS.collect()
         self.assertIn('weird_function', collected)
 
+    def test_one_of_strategy(self):
+        collected = CONFLICTING_COMMANDS.collect()
+        weird_name = collected.pop('weird_name')
+        self.assertEquals(collected, {})
+        self.assertIn(weird_name, (weird_name1, weird_name2, weird_name3))
```
Comment (markrwilliams) [inline src/gather/tests/test_commands.py]: Is it the case that the collection is deterministic, and you don't care about which function got collected in this test?  Or is it not deterministic?  #4 asks about first/last so it'd be nice to see a mention of determinism.

Also no docstrings for test methods?? 😔

File: src/gather/api.py (hunk 4 above)
Comment (moshez) [inline src/gather/api.py]: I wanted to keep the module exports to a reasonable amount, plus this function does not make a lot of sense on its own, but only as a strategy. In fact, the doc string will cover that bit :)

File: src/gather/tests/test_commands.py (hunk 8 above)
Comment (moshez) [inline src/gather/tests/test_commands.py]: Venusian only registers module-level things. The theoretically correct thing to do here is to generate ad-hoc modules, but that seemed like an overkill. The sharing should work correctly.

File: src/gather/tests/test_commands.py (hunk 9 above)
Comment (moshez) [inline src/gather/tests/test_commands.py]: Well, what's "first" and "last" if it's multiple modules? I don't want to make any guarantees, and I don't know if venusian determines an order (e.g., between releases).

Comment (moshez) [inline src/gather/tests/test_commands.py]: Re: docstring for test methods -- none of them have them, and I set pylint to disable warnings on missing docstrings. I'm going to open a separate issue to enable pylint docstring enforcement (which, by definition, will lead to docstrings).

File: src/gather/api.py [hunk 10]
```diff
@@ -37,6 +37,10 @@ def _get_modules():
         module = importlib.import_module(entry_point.module_name)
         yield module
 
+class GatherConflictError(ValueError):
+    pass
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?

File: src/gather/api.py [hunk 11]
```diff
@@ -72,25 +113,44 @@ def callback(scanner, inner_name, objct):
...
+        scanner = venusian.Scanner(update=params.update, tag=self)
         for module in _get_modules():
             scanner.scan(module, onerror=ignore_import_error)
-        return registry
+        params.raise_if_needed()
+        return params.registry
+
+@attr.s
+class _ScannerParameters(object):
+
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?  This is a nice class :)

File: src/gather/api.py [hunk 12]
```diff
@@ -72,25 +113,44 @@ def callback(scanner, inner_name, objct):
...
+    _please_raise = attr.ib(init=False, default=None)
+    _strategy = attr.ib()
+    registry = attr.ib(init=False, default=attr.Factory(dict))
+
+    def update(self, name, objct):
```
Comment (markrwilliams) [inline src/gather/api.py]: Docstring?

**Pull request merged**